File: contest/LeetCode_Weekly_Contest_201/Find_Kth_Bit_in_Nth_Binary_String.cpp
const int N = 5000005;
char buf[N];
int length;

class Solution {
public:
    void add(char* buf, int w, int s, int e) {
        for (int i = e-1; i >= s; --i) {
            int x = buf[i] - '0';
            buf[w++] = '0' + 1 - x;
        }
    }
    char findKthBit(int n, int k) {
        n--;
        k--;
        buf[0] = '0';
        length = 1;
        
        for (int i = 1; i <= n; ++i) {
            int len = length;
            buf[length++] = '1';
            add(buf, length, 0, len);
            length += len;
        }
        
        return buf[k];
    }
};

File: contest/LeetCode_Weekly_Contest_201/Find_Kth_Bit_in_Nth_Binary_String.java
class Solution {
    public char findKthBit(int n, int k) {
        ArrayList<Integer> ret = new ArrayList<Integer>();
        ret.add(0);
        n--;
        k--;

        for (int i = 0; i < n; ++i) {
            ArrayList<Integer> b = new ArrayList<Integer>();
            for (int j = ret.size()-1; j >= 0; --j) {
                b.add(1 - ret.get(j));
            }
            ret.add(1);
            ret.addAll(b);
        }

        return ret.get(k) == 1 ? '1' : '0';
    }
}
